core/storage/chain/impl/chain_store_impl.hpp

```diff
@@ -21,57 +21,85 @@
                          public std::enable_shared_from_this<ChainStoreImpl> {
    public:
-    using BlockValidator = ::fc::blockchain::block_validator::BlockValidator;
-    using WeightCalculator = ::fc::blockchain::weight::WeightCalculator;
-
     ChainStoreImpl(ChainStoreImpl &&other) = default;
 
     /* @brief creates new ChainStore instance */
     static outcome::result<std::shared_ptr<ChainStoreImpl>> create(
-        std::shared_ptr<ipfs::IpfsBlockService> block_service,
-        std::shared_ptr<ChainDataStore> data_store,
+        std::shared_ptr<IpfsDatastore> data_store,
         std::shared_ptr<BlockValidator> block_validator,
         std::shared_ptr<WeightCalculator> weight_calculator);
 
     /** @brief stores head tipset */
     outcome::result<void> writeHead(const Tipset &tipset);
 
     /** @brief loads data from block storage and initializes storage */
-    outcome::result<void> load();
+    outcome::result<void> initialize();
 
-    outcome::result<Tipset> loadTipset(const TipsetKey &key) override;
+    outcome::result<Tipset> loadTipset(const TipsetKey &key) const override;
 
-    // TODO(yuraz): FIL-151 add notifications
-    // TODO(yuraz): FIL-151 implement items caching to avoid refetching them
+    outcome::result<void> addBlock(const BlockHeader &block) override;
 
-    std::shared_ptr<ChainRandomnessProvider> createRandomnessProvider()
-        override;
+    outcome::result<Tipset> heaviestTipset() const override;
 
-    outcome::result<void> addBlock(const BlockHeader &block) override;
+    outcome::result<bool> containsTipset(const TipsetKey &key) const override;
 
-    outcome::result<BlockHeader> getBlock(const CID &cid) const override;
+    outcome::result<void> storeTipset(const Tipset &tipset) override;
 
-    outcome::result<Tipset> heaviestTipset() const override;
+    outcome::result<BlockHeader> getGenesis() const override;
+
+    outcome::result<void> writeGenesis(
+        const BlockHeader &block_header) override;
+
+    primitives::BigInt getHeaviestWeight() const override {
+      return heaviest_weight_;
+    }
+
+    /** IpfsDatastore implementation is required for using in ChainDownloader */
+    outcome::result<bool> contains(const CID &key) const override {
+      return data_store_->contains(key);
+    }
+
+    outcome::result<void> set(const CID &key, Value value) override {
+      return data_store_->set(key, value);
+    }
+
+    outcome::result<Value> get(const CID &key) const override {
+      return data_store_->get(key);
+    }
+
+    outcome::result<void> remove(const CID &key) override {
+      return data_store_->remove(key);
+    }
+
+    /** chain randomness */
+    std::shared_ptr<ChainRandomnessProvider> createRandomnessProvider()
+        override;
+
+    /** @brief head change subscription */
+    connection_t subscribeHeadChanges(
+        const std::function<HeadChangeSignature> &subscriber) override {
+      return head_change_signal_.connect(subscriber);
```

Comment:
Should send CURRENT notification?

Reply:
Yes, probably, I will add it.

Reply:
added: if chain store has current tipset, notify subscriber about curent tipset